### PR TITLE
Refine visual channel docs

### PR DESCRIPTION
**Pull request by wangqianwen0418**
- add property tables for different visual channels
- restructure doc
- fix wrong URL links

### CODE REVIEW

**Comment by wangqianwen0418** [comment]
@sehilyi , would you mind taking a look before I merge these changes?
I noticed that you are working on improving the encoding types, yet I think it is still necessary to document how to use the old version of Gosling.
I will add version control shortly so that we will have different docs for different Gosling versions.

File: docs/visual-channel.mdx [hunk 1]
```diff
@@ -4,9 +4,12 @@ title: Visual Channel
 
 import {TableWrapper} from './DocComponments/table'
 
-The visual appearance of a mark is controlled by a set of visual channels (e.g., size, position, and color). One visual channel can be bound with either a data field or just a constant value.
+The visual appearance of a mark is controlled by a set of visual channels (e.g., size, position, and color). 
+One visual channel can be either:
+- bound with a data field (as shown in line 13-16)
```
Comment (sehilyi) [inline docs/visual-channel.mdx:9]:
```suggestion
- bound to a data field (as shown in line 13-16) 
```

File: docs/visual-channel.mdx [hunk 2]
```diff
@@ -17,51 +20,32 @@ The visual appearance of a mark is controlled by a set of visual channels (e.g.,
       },
       // specify the mark type
       "mark": "line",
-      // visual channel x is bound with the data field genomic
+      
+      // visual channel x is bound with the data field "genomic"
```
Comment (sehilyi) [inline docs/visual-channel.mdx:24]:
```suggestion
      // visual channel 'x' is bound to a genomic data field
```

File: docs/visual-channel.mdx [hunk 3]
```diff
@@ -17,51 +20,32 @@ The visual appearance of a mark is controlled by a set of visual channels (e.g.,
...
+      // highlight-start
       "x": {
         "field": "position",
         "type": "genomic"
       },
-      // visual channel y is bound with the data field peak
+      // highlight-end
+      // visual channel y is bound with the data field "peak"
```
Comment (sehilyi) [inline docs/visual-channel.mdx:31]:
```suggestion
      // visual channel 'y' is bound to a quantitative data field called "peak"
```

File: docs/visual-channel.mdx [hunk 4]
```diff
@@ -17,51 +20,32 @@ The visual appearance of a mark is controlled by a set of visual channels (e.g.,
...
-| field                     | string                      | specify name of the data field                                                     |
-| type                      | string                      | specify type of the data field. support `"genomic"`, `"nominal"`, `"quantitative"` |
-| aggregate                 | string                      | support `"max"`, `"min"`, `"mean"`, `"bin"`, `"count"`                             |
-| domain                    | [number, number]\| string[] | specify values of the data field                                                   |
-| range                     | [number, number]\| string[] | specify values of the visual channel                                               |
-| value                     | string \| number            | assign a constant value to the visual channel                                      |
-
-
+To bind a visual channel with a data field, Gosling creates a mapping from the values of the data field (e.g., [gnes, gpos25, gpos50, ...]) to the values of the visual channel (e.g., position of a bar).
```
Comment (sehilyi) [inline docs/visual-channel.mdx:47]:
```suggestion
To bind a visual channel to a data field, Gosling creates a mapping from the values of the data field (e.g., `[gnes, gpos25, gpos50, ...]`) to the values of the visual channel (e.g., the position of a bar). 
```

File: docs/visual-channel.mdx [hunk 5]
```diff
@@ -17,51 +20,32 @@ The visual appearance of a mark is controlled by a set of visual channels (e.g.,
...
-| field                     | string                      | specify name of the data field                                                     |
-| type                      | string                      | specify type of the data field. support `"genomic"`, `"nominal"`, `"quantitative"` |
-| aggregate                 | string                      | support `"max"`, `"min"`, `"mean"`, `"bin"`, `"count"`                             |
-| domain                    | [number, number]\| string[] | specify values of the data field                                                   |
-| range                     | [number, number]\| string[] | specify values of the visual channel                                               |
-| value                     | string \| number            | assign a constant value to the visual channel                                      |
-
-
+To bind a visual channel with a data field, Gosling creates a mapping from the values of the data field (e.g., [gnes, gpos25, gpos50, ...]) to the values of the visual channel (e.g., position of a bar). 
+We call the values of data field **domain** and the values of the visual channel **range**.
```
Comment (sehilyi) [inline docs/visual-channel.mdx:48]:
```suggestion
We call the values of a data field **domain** and the values of a visual channel **range**.
```

File: docs/visual-channel.mdx [hunk 6]
```diff
@@ -80,58 +64,51 @@ For example, the code below creates a mapping from the data `field` "Stain" to t
...
+| [`link`](/docs/mark#link)         | [`x`](#x--xe), [`xe`](#x--xe), [`x1`](#x1-y1-x1e-y1e), [`x1e`](#x1-y1-x1e-y1e), [`color`](#color), [`opacity`](#opacity)             |
+| [`triangle`](/docs/mark#triangle) | [`x`](#x--xe), [`xe`](#x--xe), [`row`](#row), [`size`](#size), [`color`](#color), [`opacity`](#opacity)                              |
+| [`text`](/docs/mark#text)         | [`x`](#x--xe), [`xe`](#x--xe), [`row`](#row), [`color`](#color), [`opacity`](#opacity)                                               |
 
+## Supported Visual Channels
 
-## xe
-`xe` stands for the end of x axis. `xe` is usually used with `x` to specify the start position and the end position of a visual mark in the horizontal direction, respectively.
+### x & xe
+`x` specify a mark's position in the horizontal direction.
```
Comment (sehilyi) [inline docs/visual-channel.mdx:83]:
```suggestion
`x` specifies a mark's position in the horizontal direction. 
```

File: docs/visual-channel.mdx [hunk 7]
```diff
@@ -80,58 +64,51 @@ For example, the code below creates a mapping from the data `field` "Stain" to t
...
+Following the common practice in genomic visluazation, Gosling uses `x` axis inclusively for genomic coordinates. 
 
-Apart from the properties shared by all channels, `xe` channel have the following unique properties:
+<TableWrapper objName="X" />
 
-| unique properties | type   | description                                                                 |
-| ----------------- | ------ | --------------------------------------------------------------------------- |
-| aggregate         | string | support `"max"`, `"min"`, `"count"`, `"mean"`, `"bin"`                      |
-| axis              | string | specify the axis position, support "none", "top", "bottom", "left", "right" |
+`xe` stands for the end of x axis. `xe` is usually used with `x` to specify the start position and the end position of a visual mark in the horizontal direction, respectively.
```
Comment (sehilyi) [inline docs/visual-channel.mdx:88]:
```suggestion
`xe` stands for the end position of a visual mark along the x-axis. This is usually used together with `x` to specify the start and the end positions of a visual mark in the horizontal direction.
```

File: docs/visual-channel.mdx [hunk 8]
```diff
@@ -80,58 +64,51 @@ For example, the code below creates a mapping from the data `field` "Stain" to t
...
 
-## y
+### y & ye
 `y` specify a mark's position in the vertical direction.
```
Comment (sehilyi) [inline docs/visual-channel.mdx:91]:
```suggestion
`y` specifies a mark's position in the vertical direction.
```

File: docs/visual-channel.mdx [hunk 9]
```diff
@@ -80,58 +64,51 @@ For example, the code below creates a mapping from the data `field` "Stain" to t
...
-| baseline          | string \| number |                                                                             |
+<TableWrapper objName="Y" />
 
-## ye
 `ye` stands for the end of y axis. `ye` is usually used with `x` to specify the start position and the end position of a visual mark in the vertical direction, respectively.
 
-## x1 x1e y1 y1e
-The four channels are used together only in `link` mark. In this case, `x` and `xe` are used with `x1` and `x1e` to specify a pair of genomic intervals that needs to be connected using band representations. Similarly, `y` and `ye` can be used with `y1` and `y1e` to show band connection along vertical axis.
+### x1 x1e y1 y1e
+The four channels are used together only in `link` mark.
```
Comment (sehilyi) [inline docs/visual-channel.mdx:98]:
```suggestion
The four channels are used together only with `betweenLink` and `withinLink` marks.
```

File: docs/visual-channel.mdx [hunk 10]
```diff
@@ -80,58 +64,51 @@ For example, the code below creates a mapping from the data `field` "Stain" to t
...
+<TableWrapper objName="Y" />
 
-## ye
 `ye` stands for the end of y axis. `ye` is usually used with `x` to specify the start position and the end position of a visual mark in the vertical direction, respectively.
 
-## x1 x1e y1 y1e
-The four channels are used together only in `link` mark. In this case, `x` and `xe` are used with `x1` and `x1e` to specify a pair of genomic intervals that needs to be connected using band representations. Similarly, `y` and `ye` can be used with `y1` and `y1e` to show band connection along vertical axis.
+### x1 x1e y1 y1e
+The four channels are used together only in `link` mark. 
+In this case, `x` and `xe` are used with `x1` and `x1e` to specify a pair of genomic intervals that needs to be connected using band representations. Similarly, `y` and `ye` can be used with `y1` and `y1e` to show band connection along vertical axis.
```
Comment (sehilyi) [inline docs/visual-channel.mdx:99]:
```suggestion
In this case, `x` and `xe` are used with `x1` and `x1e` to specify a pair of genomic intervals that need to be connected using band representations. Similarly, `y` and `ye` can be used with `y1` and `y1e` to show band connection along the vertical axis.
```

File: docs/visual-channel.mdx [hunk 11]
```diff
@@ -80,58 +64,51 @@ For example, the code below creates a mapping from the data `field` "Stain" to t
...
 Channel `row` is used with channel `y` to stratify a visualization with categorical values.
 
-Without specifying `row`:
-
+- Without specifying `row`:
 <img src="/img/doc_images/without_row.png" width="500" alt="without row example"/>  
 
-Line charts are stratified with sample names.
-
+- Line charts are stratified with sample names (`"row": {"field": "sample"}`).
```
Comment (sehilyi) [inline docs/visual-channel.mdx:111]:
```suggestion
- Line charts are stratified with sample names (`"row": { "field": "sample", "type": "nominal" }`).
```

File: docs/visual-channel.mdx [hunk 12]
```diff
@@ -169,35 +146,40 @@ Line charts are stratified with sample names.
...
 
 `text` channel is used only in `text` mark to specify what textual information to display.
 
-## color
-Channel `color` specifies the filling color of the mark. Binding `color` with categorical values in `bar` and `area` marks stack marks that are positioned in the same genomic intervals to better show their cumulative values.
+<TableWrapper objName="Text" />
 
-Apart from the properties shared by all channels, the `color` channel have the following unique properties:
+### color
+Channel `color` specifies the filling color of the specified mark.
```
Comment (sehilyi) [inline docs/visual-channel.mdx:163]:
```suggestion
The channel `color` specifies the foreground color of a mark.
```

File: docs/visual-channel.mdx [hunk 13]
```diff
@@ -169,35 +146,40 @@ Line charts are stratified with sample names.
...
 
-## strokeWidth
+### strokeWidth
 Channel `strokeWidth` defines the outline thickness of the mark shape. Gosling supports `strokeWidth` in the following marks: `rect`, `area`, `point`, `bar`, `link`.
+<TableWrapper objName="StrokeWidth" />
 
-## opacity
-Channel `opacity` specifies the opacity of the mark shape.
+### opacity
+Channel `opacity` specifies the opacity of the specified mark.
```
Comment (sehilyi) [inline docs/visual-channel.mdx:177]:
```suggestion
The `opacity` channel determines the transparency level of a mark. The opacity of the foreground color (i.e., `color`) and the outline color (i.e., `stroke`) will be affected by this channel.
```

File: src/css/custom.css [hunk 14]
```diff
@@ -43,6 +43,20 @@ code, .prism-code {
...
+  counter-reset: line-number;
+}
+
+/* add number line */
+.prism-code.language-javascript .token-line::before {
+  counter-increment: line-number;
+  content: counter(line-number);
+  margin-right: calc(var(--ifm-pre-padding) * 1.5);
+  text-align: right;
+  min-width: 2.5rem;
```
Comment (sehilyi) [inline src/css/custom.css:55]: I really like the numbering!

I was wondering if we could reduce the width of the line number to save some space by changing this value to `1.5rem` which seems to work up to three digits.

```suggestion
  min-width: 1.5rem;
```

## After
Screen Shot 2021-11-05 at 3 15 21 PM

## Before 
Screen Shot 2021-11-05 at 3 15 57 PM

**Review by sehilyi — APPROVED** [review]
Thank you so much for this update! I really like the update on the code block appearance. I just left small suggestions, and this looks good to go.

File: docs/data.mdx [hunk 15]
```diff
@@ -107,9 +107,11 @@ This format allows users to include data directly in the Gosling's JSON specific
 }
 ```
 
-<TableWrapper objName='BIGWIGData'/>
+<TableWrapper objName='BIGWIGData' />
 
+### BAM (No HiGlass Server)
 
+<TableWrapper objName='BAMData' includeDescription={true}/>
```
[the remark this answers is not in the document]
Comment (wangqianwen0418) [reply docs/data.mdx:114]: Yes, you are correct! For example, in this case, the comments before 'BAMData' will display before its property table

File: docs/visual-channel.mdx [hunk 16]
```diff
@@ -17,51 +20,32 @@ The visual appearance of a mark is controlled by a set of visual channels (e.g.,
       },
       // specify the mark type
       "mark": "line",
-      // visual channel x is bound with the data field genomic
+      
+      // visual channel x is bound with the data field "genomic"
+      // highlight-start
```
[the remark this answers is not in the document]
Comment (wangqianwen0418) [reply docs/visual-channel.mdx:25]: Haha, thanks!

File: src/css/custom.css (hunk 14 above)
Comment (wangqianwen0418) [reply src/css/custom.css:55]: sounds great to me!

**Comment by wangqianwen0418** [comment]
Many thanks for the review!

**Pull request merged**